ApplicationInsights/Tests/ApplicationInsights.tests.ts

```diff
@@ -390,14 +453,94 @@ export class ApplicationInsightsTests extends TestClass {
                 }
         });
     }
+
+    private addTrackMetricTests() {
+        this.testCase({
+            name: 'TrackMetricTests: treackMetric batches metrics sent in a hot loop',
+            test: () => {
+                // Setup
+                const plugin = new TestPlugin();
+                var core = new AppInsightsCore();
+                core.initialize(
+                    {instrumentationKey: "key"},
+                    [plugin]
+                );
+                var appInsights = new ApplicationInsights();
+                appInsights.initialize({ "instrumentationKey": "ikey" }, core, []);
+                var trackStub = this.sandbox.stub(appInsights.core, "track");
+
+                // Act
+                appInsights.trackMetric({name: "test metric", average: 0});
+                this.clock.tick(1);
+
+                // Verify
+
+                Assert.ok(trackStub.calledOnce, "core.track was called once after sending one metric");
+                trackStub.reset();
+
+                // Act
+                for (var i = 0; i < 100; i++) {
+                    appInsights.trackMetric({name: "test metric", average: 0});
+                }
+                this.clock.tick(1);
+
+                // Test
+                Assert.equal(100, trackStub.callCount, "core.track was called 100 times for sending 100 metrics");
+            }
+        });
+    }
+
+    private getFirstResult(action: string, trackStub: SinonStub, skipSessionState?: boolean): ITelemetryItem {
+        const index: number = skipSessionState ? 1 : 0;
+
+        Assert.ok(trackStub.args && trackStub.args[index] && trackStub.args[index][0], "track was called for: " + action);
+        return <ITelemetryItem>trackStub.args[index][0];
+    }
 }
 
 class TestPlugin implements IPlugin {
-    private _config: IConfiguration;
-    priority: number = 100;
 
-    public initialize(config: IConfiguration) {
-        this._config = config;
-        // do custom one time initialization
+    public isFlushInvoked = false;
+    public isTearDownInvoked = false;
+    public isResumeInvoked = false;
+    public isPauseInvoked = false;
+
+    constructor() {
+        this.processTelemetry = this._processTelemetry.bind(this);
+    }
+    public pause(): void {
+        this.isPauseInvoked = true;
+    }    
+    
+    public resume(): void {
+        this.isResumeInvoked = true;
+    }
+
+    public teardown(): void {
+        this.isTearDownInvoked = true;
     }
-}
\ No newline at end of file
+
+    flush(async?: boolean, callBack?: () => void): void {
+        this.isFlushInvoked = true;
+        if (callBack) {
+            callBack();
+        }
+    }
+
+    public processTelemetry;
+
+    public identifier = "Sender";
+    
+    setNextPlugin(next: any) {
+        // no next setup
+    }
+
+    public priority: number = 201;
```

Comment:
change to something else like 150 for test.

Reply:
9317f51